src/api/messages/__tests__/migrateMessages-test.js

```diff
@@ -1,12 +1,49 @@
+/* @flow strict-local */
+import omit from 'lodash.omit';
+
 import { migrateMessages } from '../getMessages';
+import * as eg from '../../../__tests__/lib/exampleData';
+import type { ServerMessage, ServerReaction } from '../getMessages';
+import type { Message } from '../../modelTypes';
 
 describe('migrateMessages', () => {
-  test('Replace user object with `user_id`', () => {
-    const messages = [
-      { reactions: [{ user: { id: 1, full_name: 'name', email: 'a@a.com' }, code: 'code' }] },
-    ];
-    const expectedOutput = [{ reactions: [{ user_id: 1, code: 'code' }] }];
+  const reactingUser = eg.makeUser();
+
+  const serverReaction: ServerReaction = {
+    emoji_name: '+1',
+    reaction_type: 'unicode_emoji',
+    emoji_code: '1f44d',
+    user: {
+      email: reactingUser.email,
+      full_name: reactingUser.full_name,
+      id: reactingUser.user_id,
+    },
+  };
+
+  const serverMessage: ServerMessage = {
+    ...(omit(eg.streamMessage(), 'reactions'): $Diff<Message, { reactions: mixed }>),
```

Comment:
Is the `omit` needed here? The property mentioned after the spread should clobber it anyway.

Reply:
Definitely sounds reasonable.

It seems like Flow happily verifies that there will be a correct type at `reactions` after the spread, but only with the `omit` (or something like it) in place. Even though it's in the spec (er, I'm pretty sure) that later-listed properties will clobber earlier-listed ones, it seems like Flow maybe doesn't know that, or in any case doesn't like it when there are two competing (and both present) values for a single property.

If I take away the `omit`:

```
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/api/messages/__tests__/migrateMessages-test.js:23:40

Cannot assign object literal to serverMessage because:
 • property user is missing in object type [1] but exists in object type [2] in array element of property reactions.
 • property user_id is missing in object type [2] but exists in object type [1] in array element of property
   reactions.

     src/api/messages/__tests__/migrateMessages-test.js
      20│     },
      21│   };
      22│
      23│   const serverMessage: ServerMessage = {
      24│     ...eg.streamMessage(),
      25│     reactions: [serverReaction],
      26│   };
      27│
      28│   const input: ServerMessage[] = [serverMessage];
      29│

     src/api/messages/getMessages.js
 [2]  33│   reactions: $ReadOnlyArray<ServerReaction>,

     src/api/modelTypes.js
 [1] 493│   reactions: $ReadOnlyArray<Reaction>,
```

Reply:
Hmm, indeed.

It looks like this is one of the issues discussed here:
https://medium.com/flow-type/spreads-common-errors-fixes-9701012e9d58
> Spreads now overwrite properties instead of inferring unions

and it's fixed in Flow v0.111.

Which, checking our next RN upgrade issue #3782... we'll get with that upgrade to RN v0.62! Should be a lot of Flow improvement in that upgrade -- that'll be a happy day.

Reply:
So this is a fine workaround; but let's add a comment mentioning Flow v0.111. (So that when we do that upgrade we can easily find this in a grep and make it one of the spots where we improve the types as a followup.)